### ISSUE TITLE
Gradio finetuning on ~60k samples of audio, are these settings ok? I've some doubt

### ISSUE THREAD

**Issue by MithrilMan** [SWivid/F5-TTS#566]
### Checks

- [X] This template is only for usage issues encountered.
- [X] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [X] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Environment Details

![image](https://github.com/user-attachments/assets/f6014bd4-e587-42c9-aba2-368915d1a545)

pytorch/pytorch:2.5.1-cuda12.4-cudnn9-runtime as base template

### Steps to Reproduce

- configured the vast.ai template pytorch/pytorch:2.5.1-cuda12.4-cudnn9-runtime 
- installed stuff listed in dockerfile of this repo

### ✔️ Question

Does this setup sounds right to you?
![image](https://github.com/user-attachments/assets/65b0317f-aec4-4fb7-a8c0-bb2bcb1e5f61)

Even if I'm not a kid (not at all :D) these are my first steps into AI finetuning so I'd like to know if these settings sounds good to you.
It's not clear to me the relation between Batch size per GPU and the other settings.

By trial and error I managed to set 7000 as batch size and my GPU memory is quite full
```
/workspace/F5-TTS/ckpts$ nvidia-smi --query-gpu=memory.used,memory.free --format csv
memory.used [MiB], memory.free [MiB]
19552 MiB, 4560 MiB
22406 MiB, 1706 MiB
```
Actually my finetuning is running without OOM and it already saved 2 checkpoints so I think it shold work till the end (if I don't finish the disk, need to do some math first)

also the precision section is quite obscure to me
from my understanding, by using the 4090 it's right to chose "none" instead of any of these half precision because the 4090 is strong on fp32, correct?

I want to test out an H100 machine, in that scenario, beside finding the right amount of batch size, should I pick the 8-bit adam optimizer?

Would be nice to have some clarification about how these settings works, I tried to look for information but it's hard to have a clear view, thanks to anyone clarifying my doubts.

**Comment by SWivid** [comment]
> from my understanding, by using the 4090 it's right to chose "none" instead of any of these half precision because the 4090 is strong on fp32, correct?

fp16 better (since faster)

> 8-bit adam

you can just turn it on (will save some gpu memory)

> other settings

there are just commonly used ones, i think chatgpt and stuff can give with quite satisfactory and correct answers/explanations lol

**Comment by MithrilMan** [comment]
@SWivid what's confusing me the most is how to chose Batch size per GPU
I read somewhere in this repo that you have to do "BatchSizePerGPU * 256 * 24000" to get the amount of used memory, but it's not clear to me how that batch is feeded:
did you fit into that buffer the audio of all the samples (capped at MaxSamples value)?

if no, how does it works?

if yes what if a sample doesn't fit in the frame, is the whole sample dropped? because if it's just taken partially, the reference text could be a problem because could not correspond to the audio segment, while if it's not taken, what happen if an audio sample doesn't fit in the BatchSizePerGPU if it is too small?

I mean, you can see how confuse I am :D (and looking into chatgpt hallucination wouldn't help me that much lol)

**Comment by SWivid** [comment]
> if yes what if a sample doesn't fit in the frame, is the whole sample dropped?

data is first be organized with the dataset class where you could do filtering here (or when you setup your original dataset)
https://github.com/SWivid/F5-TTS/blob/eea65de823daa37c9d5b565cc01254d0c94bcc58/src/f5_tts/model/dataset.py#L136-L138

> BatchSizePerGPU

if you choose `batch_size_type: frame`, it means on each gpu, the max frames is batch_size_per_gpu
if you choose `batch_size_type: sample`, it is the numbers of samples

e.g. if you set 9375 frames per gpu
it means you can have 100 seconds in total (10*10seconds  audio samples in 24khz or 5*20), 100*24000/256=9375 (256 is hop_length)

**Comment by AsmoKoskinen** [comment]
> there are just commonly used ones, i think chatgpt and stuff can give with quite satisfactory and correct answers/explanations lol

Yes, I do this ;-)

Claude is very helpful, you can show tensorboard pictures and ask how to edit the configuration file. I'm still trying to understand a bit about this concept before I start recording my own voice. This is what Claude suggests when I test train a combined Common Voice FI and Vox Populi Fi dataset.

{
    "exp_name": "F5TTS_Base",
    "learning_rate": 2e-06,        // Lower initial LR for more stability
    "batch_size_per_gpu": 2000,    // Current size seems good
    "batch_size_type": "frame",
    "max_samples": 64,
    "grad_accumulation_steps": 4,   // Good for current setup
    "max_grad_norm": 0.5,          // Working well for gradient stability
    "epochs": 100,
    "num_warmup_updates": 1000,    // Increase warmup period
    "save_per_updates": 5000,      // More frequent checkpoints
    "last_per_steps": 2500,        // More frequent evaluations
    "finetune": true,
    "file_checkpoint_train": "",
    "tokenizer_type": "pinyin",
    "tokenizer_file": "",
    "mixed_precision": "fp16",      // Keep this for efficiency
    "logger": "tensorboard",
    "bnb_optimizer": true          // Enable for better memory efficiency
}

Of course, I am blind to these suggestions because I don't have the necessary education to really understand the math in all this, but this is probably the direction we are going in: training AI models with AI.

But once I roughly understand something, I can move this from my home computer to the cloud.

**Comment by SWivid** [comment]
> {
> "exp_name": "F5TTS_Base",
> "learning_rate": 2e-06, // Lower initial LR for more stability
> "batch_size_per_gpu": 2000, // Current size seems good
> "batch_size_type": "frame",
> "max_samples": 64,
> "grad_accumulation_steps": 4, // Good for current setup
> "max_grad_norm": 0.5, // Working well for gradient stability
> "epochs": 100,
> "num_warmup_updates": 1000, // Increase warmup period
> "save_per_updates": 5000, // More frequent checkpoints
> "last_per_steps": 2500, // More frequent evaluations
> "finetune": true,
> "file_checkpoint_train": "",
> "tokenizer_type": "pinyin",
> "tokenizer_file": "",
> "mixed_precision": "fp16", // Keep this for efficiency
> "logger": "tensorboard",
> "bnb_optimizer": true // Enable for better memory efficiency
> }

it's good~

p.s. if no very large storage on cloud, could set a larger interval for "save_per_updates"

not sure about learning_rate (lr) 2e-6 (we have not try with this setting)
yes a small initial lr is good (so generally we have a warmup)

here the hyperparameter is for the peak
e.g. if we set a lr of 1e-4, and warmup with 1000 updates, the lr will start from a very small one and goes up gradually to 1e-4, and then decay
so if set a lr of 2e-6, the peak with be 2e-6; if want warmup more slowly, set a larger number to num_warmup_updates

**Comment by MithrilMan** [comment]
I'm actually experimenting with 8x4090 on runpod
dataset => [ylacombe/cml-tts](https://huggingface.co/datasets/ylacombe/cml-tts)
Italian train split (~73 hours)

```
{
    "exp_name": "F5TTS_Base",
    "learning_rate": 1e-05,
    "batch_size_per_gpu": 10000,
    "batch_size_type": "frame",
    "max_samples": 64,
    "grad_accumulation_steps": 1,
    "max_grad_norm": 1,
    "epochs": 200,
    "num_warmup_updates": 2000,
    "save_per_updates": 600,
    "last_per_steps": 300,
    "finetune": true,
    "file_checkpoint_train": "",
    "tokenizer_type": "char",
    "tokenizer_file": "",
    "mixed_precision": "fp16",
    "logger": "wandb",
    "bnb_optimizer": false
}
```

after 100 epoch (using ema), the accent is still wrong, not sure if num_warmup_updates is correct (I don't understand how to cherrypick it)
also what happens if I stop and restart the training, is warmup something that work on each step or what? I saw in the paper you told that warmup linearly decrease during training, is it the same using the gradio interface?


@SWivid 
One thing I noticed on my previous attempt with around 300h, is that after a training of aroun 70 epoch the model lost the ability to clone the reference voice properly (see readme in https://huggingface.co/alien79/f5-ita-test for some detail and you can try to use a .safetensor there to see how the cloned voice fails to be cloned properly)

Note that on that attempt, my transcription was without punctuaction so lost even some ability to pause properly (while the current train dataset has puntuaction so will be better on that)

I don't understand what's the sweet spot to fine tune in order to not lose the ability to mimic the reference voice during inference.

If I test the current trained model without EMA, the pronounce is quite good already but it can't mimic the infer voice (probably because to do that we'd need way more speakers in the dataset?)

While if I let it train more, in order to let have the model to have a good pronounce even with EMA enabled, I fear I'll end again in a model that lost part of the ability to clone a reference voice.

I didn't find any good advice in other threads, do you have any reccomendation?

**Comment by SWivid** [comment]
@MithrilMan hi, you could mix some English or other language dataset (introduce more diversity to keep zero-shot ability, also samely for punctuation)

for gpu mem usage, cuz attention is quadratic to sample length (you could dynamically scale down threshold for longer samples here https://github.com/SWivid/F5-TTS/blob/8898d05e374bcb8d3fc0b1286037e95df61f491f/src/f5_tts/model/dataset.py#L196; or just not use mem fully is acceptable cuz utilization is full)
